libraries/aws/jobs/src/jobs.c

```diff
@@ -0,0 +1,565 @@
+      ( thingNameLength <= JOBS_THINGNAME_MAX_LENGTH ) )
+
+#define checkCommonParams()                    \
+    ( ( buffer != NULL ) && ( length > 0U ) && \
+      checkThingParams() && ( outLength != NULL ) )
+
+/**
+ * See jobs.h for docs.
+ *
+ * @brief Populate a topic string for a subscription request.
+ */
+JobsStatus_t Jobs_GetTopic( char * buffer,
+                            size_t length,
+                            const char * thingName,
+                            uint16_t thingNameLength,
+                            JobsTopic_t api,
+                            size_t * outLength )
+{
+    JobsStatus_t ret = JobsBadParameter;
+    size_t start = 0U;
+
+    if( checkCommonParams() &&
+        ( api > JobsInvalidTopic ) && ( api < JobsMaxTopic ) )
+    {
+        writePreamble( buffer, &start, length, thingName, thingNameLength );
+
+        if( api >= JobsDescribeSuccess )
+        {
+            ( void ) strnAppend( buffer, &start, length,
+                                 "+/", ( sizeof( "+/" ) - 1U ) );
+        }
+
+        ret = strnAppend( buffer, &start, length,
+                          apiTopic[ api ], apiTopicLength[ api ] );
+
+        if( start == length )
+        {
+            start--;
+        }
+
+        buffer[ start ] = '\0';
+        *outLength = start;
+    }
+
+    return ret;
+}
+
+/**
+ * @brief Compare the leading n bytes of two character sequences.
+ *
+ * @param[in] a  first character sequence
+ * @param[in] b  second character sequence
+ * @param[in] n  number of bytes
+ *
+ * @return JobsSuccess if the sequences are the same;
+ * JobsNoMatch otherwise
+ */
+static JobsStatus_t strnEq( const char * a,
+                            const char * b,
+                            size_t n )
+{
+    size_t i;
+
+    for( i = 0U; i < n; i++ )
+    {
+        if( a[ i ] != b[ i ] )
+        {
+            break;
+        }
+    }
+
+    return ( i == n ) ? JobsSuccess : JobsNoMatch;
+}
+
+/**
+ * @brief Wrap strnEq() with a check to compare two lengths.
+ */
+static JobsStatus_t strnnEq( const char * a,
+                             size_t aLength,
+                             const char * b,
+                             size_t bLength )
+{
+    JobsStatus_t ret = JobsNoMatch;
+
+    if( aLength == bLength )
+    {
+        ret = strnEq( a, b, aLength );
+    }
+
+    return ret;
+}
+
+/**
+ * @brief Predicate returns true for a valid job ID character
+ */
+static bool_ isJobIdChar( char a )
+{
+    bool_ ret;
+
+    if( ( a == '-' ) || ( a == '_' ) )
+    {
+        ret = true;
+    }
+    else if( ( a >= '0' ) && ( a <= '9' ) )
+    {
+        ret = true;
+    }
+    else if( ( a >= 'A' ) && ( a <= 'Z' ) )
+    {
+        ret = true;
+    }
+    else if( ( a >= 'a' ) && ( a <= 'z' ) )
+    {
+        ret = true;
+    }
+    else
+    {
+        ret = false;
+    }
+
+    return ret;
+}
+
+/**
+ * @brief Predicate returns true for a valid job ID string
+ */
+static bool_ isValidJobId( const char * jobId,
```

Comment:
Most of the static functions in this file have more detailed doxygen comments with `@param` and `@return` tags, except for the above three. Can you add those here as well?

Reply:
added